### PR TITLE
tests : add tests for shap/plots/_group_difference.py

### PR DESCRIPTION
**Pull request by patilpratik1905**
## Overview

Part of Issue #3690   

### **What was done**

This PR extends `test_group_difference.py` by adding **20 behavioural tests** on top of the existing image-based test
The new tests cover:

* Core functionality (no errors, correct return behavior)
* 1-D and 2-D input handling
* Feature name generation (auto + custom)
* Parameters like `xlabel`, `max_display`, `xmin`, `xmax`
* Figure handling (`ax=None` vs provided axes)
* Correctness of computed differences (sign and zero cases)
* Spine visibility (via parametrized test)
* Public API access (`shap.plots.group_difference`)

---

### **Why it was done**

Previously, `_group_difference.py` had very limited test coverage (~9%) and relied only on a visual regression test.
This meant key logic parameter handling, edge cases, and correctness was not being validated.

---

### **How it was tested**

Tests were run locally using:

```bash
pytest tests/plots/test_group_difference.py -v --tb=short 
```

All tests pass successfully

---

### **AI Usage Disclosure**

Kiro AI was used for initial test scaffolding, identifying uncovered branches, and structuring tests.

Manually verified:

* All assertions map to actual logic in `_group_difference.py`
* Edge cases use controlled, deterministic inputs
* Repository conventions (naming, fixtures, docstrings) are followed

---

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: tests/plots/test_group_difference.py [hunk 1]
```diff
@@ -17,3 +18,206 @@ def test_group_difference(explainer):
...
+    fig, ax = plt.subplots()
+    group_difference(shap_values, group_mask, feature_names=feature_names, show=False, ax=ax)
+    assert len(ax.patches) > 0
+
+
+def test_1d_shap_values_does_not_raise(basic_1d_data):
+    """Check that a 1-D shap_values array (single model output) is handled without error."""
+    shap_values, group_mask = basic_1d_data
+    fig, ax = plt.subplots()
+    group_difference(shap_values, group_mask, show=False, ax=ax)
```
Comment (CloseChoice) [inline tests/plots/test_group_difference.py]: no assert?

File: tests/plots/test_group_difference.py [hunk 2]
```diff
@@ -17,3 +18,206 @@ def test_group_difference(explainer):
...
+def test_1d_shap_values_auto_feature_name(basic_1d_data):
+    """Check that 1-D input with feature_names=None produces a single empty-string tick label.
+
+    The source sets feature_names = [""] for the 1-D branch when no names are supplied.
+    """
+    shap_values, group_mask = basic_1d_data
+    fig, ax = plt.subplots()
+    group_difference(shap_values, group_mask, feature_names=None, show=False, ax=ax)
+    tick_labels = [t.get_text() for t in ax.get_yticklabels()]
+    assert "" in tick_labels
```
Comment (CloseChoice) [inline tests/plots/test_group_difference.py]: what exactly are you testing here? I am unsure if this is really the desired behaviour but we won't touch functionality in this PR, I would rather have this as an mpl test.

**Review by CloseChoice — CHANGES_REQUESTED** [review]
Thanks for the PR. Could you pls trim down the changes to just a handful of tests and use pytest-mpl to generate real plots here?

**Comment by patilpratik1905** [comment]
Thank you for your review @CloseChoice 
I will make the required changes soon.

File: shap/_serializable.py [hunk 3]
```diff
@@ -90,7 +90,7 @@ def save(self, name, value, encoder="auto"):
                 pickle.dump("cloudpickle.dump", self.out_stream)
                 cloudpickle.dump(value, self.out_stream)
         elif encoder == "auto":
-            if isinstance(value, (int, float, str)):
+            if isinstance(value, int | float | str):
```
Comment (CloseChoice) [inline shap/_serializable.py:93]: pls revert all of these changes, there is no need for this and doing 2 things in one PR is not good style IMO.

File: tests/plots/baseline/test_group_difference.png
Comment (CloseChoice) [inline tests/plots/baseline/test_group_difference.png:1]: why did this baseline change? Can you elaborate?

**Review by CloseChoice — CHANGES_REQUESTED** [review]
Thanks for the PR. We need a couple of changes here till round number 2 of reviews.
